pyiron/atomistics/structure/neighbors.py

```diff
@@ -21,6 +21,27 @@
 
 s = Settings()
 
+def get_average_of_unique_labels(labels, values):
```

Comment:
I find this version

```python
return np.array([ values[ labels == s ].mean(axis = 0)
                                  for s in np.unique(labels) ])
```

much clearer, but it does construct a mostly empty index array for every label that your version avoids with the sparse matrix.  I did some mini-benchmarks and the sparse version is faster above ~10,000 values, but only by ~10% (or <1µs per value).  

No hard feelings either way, if you want to keep this version that's also fine by me.

Reply:
Oh, I agree that in things I'd show to other people I wouldn't do the same - I agree entirely with you that that's how it should be written in Notebooks.

On the other hand, as long as it's clear what the function does, and if it just does one single thing, I don't really mind having complex functions.

And here's the result of my speed comparison (beware of the log scales):

<img width="393" alt="Unknown" src="https://user-images.githubusercontent.com/37879103/99940779-2461b380-2d6d-11eb-94a6-e15cca22763e.png">

After having seen this, I might still prefer the current implementation, also because we might find applications in other areas whenever we do clustering.

Reply:
P.S. I'm gladly overruled if you still want me to change it.

Reply:
I agree, but playing around some more I found this version

```python
def get_average_of_unique_labels(labels, values):
    u, counts = np.unique(labels, return_counts = True)
    means = np.zeros( (len(u), *values.shape[1:]) )
    means[labels] += values
    #manually broadcast counts to have the same dimension as means
    means /= np.reshape(counts, (-1,) + (1,)*len(values.shape[1:]))
    return means
```

which is twice as fast than the sparse version.  Except for the reshaping, I find it quite readable too.

Reply:
Here, I don't get the same results as the other ones..

Reply:
Mh, true. `means[labels] += values` doesn't actually do what I though it would, my bad.

Reply:
Turns out you have to use `np.add.at`, but it's way too slow.  I'm ok with the sparse implementation then, but I think the output should have the same shape as values you pass in.  Right now the returned means are flattened and that might not be what you expect when you pass in e.g. matrices or other multidimensional stuff.

Reply:
Return values can have the shape of the labels, but since we don't consider multidimensional labels, I would say it makes the most sense if we flatten the array. One option I see here is to return an array of single elements, instead of returning a flattened array, when there's only one element in each component. I didn't consider this as a particularly important case, but do you want me to change this?